display_list/geometry/dl_region.cc

```diff
@@ -8,31 +8,527 @@
+  auto* dst = spans_ + res + 1;
+  memmove(dst, begin, chunk_size * sizeof(Span));
+
+  return res;
+}
+
+size_t DlRegion::SpanBuffer::getChunkSize(SpanChunkHandle handle) const {
+  FML_DCHECK(handle < size_);
+  return spans_[handle].left;
+}
+
+void DlRegion::SpanBuffer::getSpans(SpanChunkHandle handle,
+                                    const DlRegion::Span*& begin,
+                                    const DlRegion::Span*& end) const {
+  FML_DCHECK(handle < size_);
+  auto& info = spans_[handle];
+  begin = spans_ + handle + 1;
+  end = begin + info.left;
+}
+
+DlRegion::DlRegion(const std::vector<SkIRect>& rects) {
+  addRects(rects);
+}
+
+DlRegion::DlRegion() {}
+
+bool DlRegion::spansEqual(SpanLine& line,
+                          const Span* begin,
+                          const Span* end) const {
+  const Span *our_begin, *our_end;
+  span_buffer_.getSpans(line.chunk_handle, our_begin, our_end);
+  size_t our_size = our_end - our_begin;
+  size_t their_size = end - begin;
+  if (our_size != their_size) {
+    return false;
+  }
+
+  return memcmp(our_begin, begin, our_size * sizeof(Span)) == 0;
+}
+
+DlRegion::SpanLine DlRegion::makeLine(int32_t top,
+                                      int32_t bottom,
+                                      const SpanVec& v) {
+  return makeLine(top, bottom, v.data(), v.data() + v.size());
+}
+
+DlRegion::SpanLine DlRegion::makeLine(int32_t top,
+                                      int32_t bottom,
+                                      const Span* begin,
+                                      const Span* end) {
+  auto handle = span_buffer_.storeChunk(begin, end);
+  return {top, bottom, handle};
+}
+
+size_t DlRegion::unionLineSpans(std::vector<Span>& res,
+                                const SpanBuffer& a_buffer,
+                                SpanChunkHandle a_handle,
+                                const SpanBuffer& b_buffer,
+                                SpanChunkHandle b_handle) {
+  const Span *begin1, *end1;
+  a_buffer.getSpans(a_handle, begin1, end1);
+
+  const Span *begin2, *end2;
+  b_buffer.getSpans(b_handle, begin2, end2);
+
+  size_t min_size = (end1 - begin1) + (end2 - begin2);
+  if (res.size() < min_size) {
+    res.resize(min_size);
+  }
+
+  // Pointer to the next span to be written.
+  Span* new_span = res.data();
+
+  while (true) {
+    if (begin1->right < begin2->left) {
+      *new_span++ = *begin1++;
+      if (begin1 == end1) {
+        break;
+      }
+    } else if (begin2->right < begin1->left) {
+      *new_span++ = *begin2++;
+      if (begin2 == end2) {
+        break;
+      }
+    } else {
+      break;
+    }
+  }
+
+  Span current_span{0, 0};
+  while (begin1 != end1 && begin2 != end2) {
+    if (current_span.left == current_span.right) {
+      if (begin1->right < begin2->left) {
+        *new_span++ = *begin1++;
+      } else if (begin2->right < begin1->left) {
+        *new_span++ = *begin2++;
+      } else if (begin1->left < begin2->left) {
```

Comment:
Is it safe to say at this point that begin1 and begin2 overlap? Could this be simplified to:

```
  current_span = {
      std::min(begin1->left, begin2->left),
      std::max(begin1->right, begin2->right)
  };
  begin1++;
  begin2++;
```

Reply:
Yes, this should be safe to do.